In-context document support for Anthropic and Google models

alex-torregrosa opened the pull request:
## Summary

Added support for sending full documents in the chat context for Anthropic and Google endpoints. 

New available types:
* Anthropic: PDF files
* Google: text documents, PDF files, audio files

Limitations:
* Not working with Agents, as langchain does not support pdf files.
* Token count prediction with anthropic is wrong, as with the current dependencies we can't get the PDF page count.

Feel free to suggest any changes or anything else that might be needed.

## Change Type

- [x] New feature (non-breaking change which adds functionality)


## Testing

Tested both the Anthropic & Google endpoints uploading PDF files and requesting a summary. Same with audio files on the Google endpoint.

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] My changes do not introduce new warnings

alex-torregrosa committed:
Upload files directly to models when RAG is disabled

Reused the image handling functions, as the mime type was already being
set correctly.

For now the upload only happens if RAG is disabled. A better approach
would be to re-use the agents menu for deciding context vs RAG

alex-torregrosa committed:
Moved image encoding functions to upper `Files` directory

Will be generalized for any file type, not just images

alex-torregrosa committed:
Rename addImageURLs to addFileURLs

alex-torregrosa committed:
Cleanup file-related functions previously used only for images

Removed as much references to images as possible

alex-torregrosa committed:
Renamed encode.js image_urls to file_urls

alex-torregrosa committed:
Filter allowed mime types for Anthropic and Google, add audio

Sets mime type filters for the Google and Anthropic endpoints according
to the documents that can be attached in the chat context. Also brings
support to upload audio files directly to Gemini.

alex-torregrosa committed:
PDF support for claude sonnet 3.5

Added support for attaching documents to claude 3.5

alex-torregrosa commented:
Ok! I can update the MR to do something similar reusing the `AttachFileMenu` for non-agent endpoints.

Maybe something like this? (with the right icon)
![image](https://github.com/user-attachments/assets/d5573540-7525-456d-8030-3a74f8c3ceef)

On the backend side, the requests could be differentiated by marking RAG files with `EToolResources.file_search`.

I'm happy to start working on these changes, or if you'd prefer to handle it, that's fine as well. Let me know what you'd like to do.

danny-avila commented:
@alex-torregrosa thank you that looks good, only small nitpick maybe "Upload to Provider" makes more sense.

> On the backend side, the requests could be differentiated by marking RAG files with EToolResources.file_search.

Sounds good to me. I'm happy to let you work on it. 

Note that I was planning on migrating all the file options to every non-agents endpoint, and then start using "ephemeral" agents when certain options/features are selected. The plan is to eventually migrate all backend chat operations to agents, as there is improved performance there and extends capabilities for tool use (which is often assumed as it is for ChatGPT, as is the motivation for this change).

I can work on that after your changes, though, as I can see that being iterative over the current scope.

alex-torregrosa committed:
Allow switching between RAG and native file input in GUI

AttachFileMenu is now used for all endpoints, and RAG file uploads are
sent with the `file_search` tool.

Based on the tool, the upload handler can now select between vectorDB or
local files, without need to check for an empty `RAG_API_URL`.

alex-torregrosa commented:
Done!

![image](https://github.com/user-attachments/assets/2185ffa4-f5ff-43d8-b4e0-483e1c939e7c)

Now the menu will show "Upload to provider" for providers that support more file types than images, and "Upload image" otherwise:
![image](https://github.com/user-attachments/assets/04b7f5e8-ed20-424a-aab0-44f745fe9a5a)

Some rough file filtering is also done per-model, similar to the old `image/*` one.

raphaelgurtner commented:
This looks great and what I had hoped for for some time now, any specific reason why it isn't merged yet? Do you need help with something?

danny-avila commented:
This PR needs a bit more work on my end. It will be merged within the next 2 weeks, see:
https://www.librechat.ai/blog/2025-02-20_2025_roadmap#upcoming-features

marcelamsler commented:
I would really appreciate if this is merged as this feature would add so much value for all that use Vertex-AI with Antrophic and Gemini Models

danny-avila commented:
> I would really appreciate if this is merged as this feature would add so much value for all that use Vertex-AI with Antrophic and Gemini Models

It's on the roadmap, and not ready to be merged

Pfaufisch commented:
@danny-avila Any update on this?

danny-avila commented:
> @danny-avila Any update on this?

It's moved to v0.7.9 release

thelinuxkid commented:
> > @danny-avila Any update on this?
> 
> It's moved to v0.7.9 release

Thanks Danny. Just curious if there's a date planned yet for the v0.7.9 release, just want align our team accordingly.

marcelo-mascarenhas commented:
My team is planning to adopt LibreChat, and being able to upload documents directly to the provider (without relying on RAG initially) is important for our use case in this first stage. If this feature isn’t available soon, we’re considering adapting the code ourselves to meet this requirement. Any updates?

Pfaufisch commented:
Hi @danny-avila, another update on this feature would be very much appreciated! It looks like it's not going to be included in v0.7.9?

(I do not want to sound too ungrateful, I see and appreciate the amount of work you're putting into LibreChat! It's just that seeing this PR becoming stale, despite the feature being announced since a couple of months, is causing a bit of frustration on my end :))

thelinuxkid commented:
I've create a PR with a working solution for this here https://github.com/danny-avila/LibreChat/pull/8973

thelinuxkid commented:
Thanks!

raphaelgurtner commented:
🥳 FYI this has been implemented here (separate PRs for each provider) https://github.com/danny-avila/LibreChat/pull/9088
